### PR TITLE
Standardize on cargo-c for building rustls-ffi, CMake for building test programs

### PR DESCRIPTION
**Pull request by cpu**
I recommend reviewing this branch commit-by-commit. There are two main goals:

1. Standardizing [cargo-c](https://github.com/lu-zero/cargo-c) for building a `rustls-ffi` library to use as a consumer.
2. Standardizing [cmake](https://cmake.org/) for building the `client`/`server` examples, and for driving dev. tools (formatting/running integration tests/etc).

## Standardizing on cargo-c

We introduced support for building a static or dynamic `rustls-ffi` library using cargo-c back in https://github.com/rustls/rustls-ffi/pull/274. Since then it's been used successfully by several of our downstream packagers ([archlinux](https://archlinux.org/packages/extra/x86_64/librustls/), [homebrew](https://formulae.brew.sh/formula/rustls-ffi), [gentoo](https://packages.gentoo.org/packages/net-libs/rustls-ffi), [nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/ru/rustls-ffi/package.nix), etc). We're also seeing downstream projects like curl [working around missing .pc files](https://github.com/curl/curl/pull/15193) for the `Makefile` build.

I propose we standardize on `cargo c` as the preferred way to build and install the library both in our docs and CI.  It does an excellent job of providing a convenient way to both build and _install_ the library for both static and dynamic linking, with pkg-config consideration, on all of our supported OSes, without needing to get into the nitty gritty of `rustc`. We benefit from being able to focus on `rustls-ffi` instead of the minutiae of building native libraries in a variety of tricky contexts & downstream distributors and users can benefit from a consistent approach that matches other Rust software using `cargo-c`.

Specifically, `cargo-c`:

* Is very easy to use, and avoids the need to specify flags like `--crate-type=cdylib` or to remember to install the header file manually.
* Frees us from having to maintain a manual `Makefile` installation process that didn't support Windows.
* Supports both dynamic linking and static linking, where our `Makefile` only supported the latter.
* Supports [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/) out of the box.
* Is easy to install (via [system package](https://github.com/lu-zero/cargo-c?tab=readme-ov-file#availability), `cargo`, or downloading pre-built binaries as we do in CI).

Notably standardizing on `cargo c` does **not** require it. Adventurous users are free to find the right incantations to build the library with standard `rust` tooling. The only caveat is that they're on their own for determining the right arguments for their platform/linking type, installing the header file, generating .pc files, etc etc, but the library will build without `cargo c`.

## Standardizing on CMake

Previously the GNU `Makefile` had three roles for Linux/MacOS systems:

* building a `rustls-ffi.a` static library, and installing it and the `rustls.h` header file into the correct location (_proposed to be replaced by `cargo-c`, per above_).
* building our `client` and `server` example C applications.
* offering helpful shortcuts for development tasks (code formatting, running tests, etc)

In parallel we maintained a CMake based system in order to provide a better integration story for Windows and MSVC toolchains, but did not offer the development task helpers.

In addition to the duplication of work, maintaining the `Makefile` feels like it's starting to hit a complexity wall. We're shimming in new feature flags as environment variables but there's no good way for a consumer to list the existing options or intuit their values without reading the makefile code. It's also difficult to make complex logical expressions; instead you must [nest each conditional](https://stackoverflow.com/a/6476553). In practice I've also found IDE integration poor: CLion does very well with CMake/Rust projects and not so well with GNU Makefile projects. We could reach for something like GNU autotools, but this is the path to madness and we're still stuck with CMake for Windows.

I propose we bite the bullet and standardize on `CMake`. It's a mildly unpleasant build system with a bit of a learning curve but we're stuck with it for Windows and so we might as well make the most of its cross-platform abilities. The extra power is beneficial for expressing our build options and more complex logic like ("FIPS and dynamic linking is ok, but only on UNIX"). It's also helpful having a standardized way to list existing options/descriptions for a better user experience.

Notably `cmake` is only required for building the **example** binaries, where you also need a C compiler. In this context it feels very reasonable to request `cmake` as a prerequisite. Users only interested in the `rustls-ffi` library only require `cargo`, and `cargo-c`.

Resolves https://github.com/rustls/rustls-ffi/issues/390 where there's some additional context/discussion.

### CODE REVIEW

**Comment by cpu** [comment]
@ctz I'm curious if you have any strong feelings about this before I try and get Jsha's eyes on the proposal. WDYT?

**Review by ctz — COMMENTED** [review]
This looks good & reasonable to me, and I feel I know cmake a bit better after reading the commit messages 🥴

File: cmake/rust.cmake
Comment (cpu) [inline cmake/rust.cmake:16]: one thing to improve here: we're building too much.

Previously the Windows CMake had `BUILD_ALWAYS true` with the associated comment, but used `cargo` as a `COMMAND` and left the `BUILD_COMMAND` empty. 

I don't think the choice of `COMMAND` vs `BUILD_COMMAND` is load bearing here (?) but I think `cargo capi` is building even when there aren't changes in the Rust code. Combined with `BUILD_ALWAYS true` this means we build everything for each target which is a bummer.

I'll give this some thought.

**Comment by cpu** [comment]
@divergentdave is there any chance you'd be interested in taking a look at this diff as the author of the original `CMakeLists.txt`? :sweat_smile:

**Comment by divergentdave** [comment]
Sure! I can take a look

**Comment by cpu** [comment]
> I need to debug dynamic linking in CI on Windows.

I was able to reproduce the issue I'm seeing in a [small standalone repo](https://github.com/cpu/example-ffi) and [asked the cargo-c folks](https://github.com/lu-zero/cargo-c/issues/426) to weigh in. There's a strong chance I'm putting square pegs in round holes, win32 is greek to me.

**Comment by cpu** [comment]
> I need to debug dynamic linking in CI on Windows.

All fixed w/ https://github.com/lu-zero/cargo-c/pull/428. I'm building cargo-c from source on Windows to fix CI (at the cost of runtime) but will adjust it to use a binary release when available.

We now have working CI for building the example binaries using static **and** dynamic linking for all of Linux, MacOS and Windows :tada:

File: .gitignore
Comment (divergentdave) [inline .gitignore:5]: FYI, the commit message changing this mentioned VS Code, but that uses `.vscode` instead.

**Review by divergentdave — APPROVED** [review]

File: cmake/rust.cmake
Comment (cpu) [reply cmake/rust.cmake:16]: > Or maybe cargo-c and an IDE were fighting over one hash key in the target directory?

I think that's a likely culprit. I haven't been able to reproduce this locally once I closed CLion and made sure my `cargo-c` install was up to date. I was mostly building on CLI but had also set up the IDE integration and suspect that's a recipe for rebuilds.

Thanks for checking!

File: README.md
Comment (cpu) [reply README.md:247]: Good catch! Thank you. 

> Presumably we need an explicit --debug instead of an explicit --release there.

I switched to using `"$<IF:$<CONFIG:Release>,--release,--debug>"` so that we're explicit in both cases and don't have to worry about upstream defaults. 

I was able to reproduce the problem you ran into and this fixed it for me locally.

File: cmake/options.cmake
Comment (cpu) [reply cmake/options.cmake]: You're right, thanks! Fixed up.

File: .github/workflows/test.yaml
Comment (cpu) [reply .github/workflows/test.yaml:15]: We usually administratively merge PRs that change the job names and then fix the rules up afterwards. I'll do the same here.

File: cmake/rust.cmake
Comment (cpu) [reply cmake/rust.cmake:28]: Good data point, thank you. I wasn't sure the stdout redirect would work on Windows but it sounds like your attempt to repro might have bailed before that point :thinking:

File: cmake/rust.cmake
Comment (cpu) [reply cmake/rust.cmake]: Oh that's a much better solution. Thank you for sharing :stars: 

I removed this manual handling in favour of using that generator expression var.

File: cmake/format.cmake
Comment (cpu) [reply cmake/format.cmake]: Seems reasonable. I went ahead and did that in a separate commit in case anyone thinks it's more clunky than valuable to duplicate the function defns.

File: .gitignore
Comment (cpu) [reply .gitignore:5]: Argh thanks! Fixed up.

As a Windows-outsider it boggles the mind that there are two, entirely separate, IDEs both practically named "Visual Studio" but they aren't replacements for one another!

For a brief period I was going crazy trying to figure out how to use `dumpbin` from a powershell prompt in VS Code until I realized I had to use Visual Studio (_not Code_) to launch a special developer powershell that had its path preconfigured for a bunch of useful dev tools. :dizzy_face:

**Comment by cpu** [comment]
@divergentdave Thank you for the thorough review. I really appreciate it.

File: README.md
Comment (cpu) [reply README.md:247]: _(I also fixed the mistaken references to a `integration_test` target - it's `integration-test`)._

File: cmake/rust.cmake
Comment (divergentdave) [reply cmake/rust.cmake:28]: I think stdout redirection with > should be safe, it's supported by cmd.exe.

**Comment by cpu** [comment]
> This looks good & reasonable to me, and I feel I know cmake a bit better after reading the commit messages 🥴

@ctz Would you be willing to turn this comment into a +1 review? I emailed Jsha Nov 27th to ask for his review and will give him a bit more time before I merge but in the meantime it'd be helpful to have your approval recorded.

**Review by ctz — APPROVED** [review]

File: README.md
Comment (jsha) [inline README.md:47]:
```suggestion
To change where the library is installed, use `--prefix` like:
```

File: README.md
Comment (jsha) [inline README.md:54]: I think `.dll` should be on this list now too, right?

**Review by jsha — APPROVED** [review]
This is a really nice and extensive cleanup. Thanks for working on it!

File: README.md
Comment (cpu) [reply README.md]: Good catch, thank you. Fixed up.

File: README.md
Comment (cpu) [reply README.md]: Good point. I added `.dll` and `.lib` to cover both dynamic and static linking on Windows.

**Comment by cpu** [comment]
> 35 successful and 19 expected checks

Admin merging to bypass the stale rules (which I will fix afterwards).

**Comment by cpu** [comment]
> which I will fix afterwards

Done.

**Pull request merged**
